DirectProgramming/DPC++FPGA/Tutorials/Features/experimental/ip_authoring_mmhost/src/mmhost.cpp

```diff
@@ -0,0 +1,133 @@
+#include <iomanip>
+#include <sycl/ext/intel/experimental/interfaces.hpp>
+#include <sycl/ext/intel/fpga_extensions.hpp>
+
+#include "dpc_common.hpp"
+
+using namespace cl::sycl;
+using ext::intel::experimental::property::usm::buffer_location;
+
+template <typename T>
+inline std::shared_ptr<T> make_malloc_shared(queue &q, int n, int kBL) {
+  T *mem =
+      malloc_shared<T>(sizeof(T) * n, q, property_list{buffer_location(kBL)});
+  return std::move(
+      std::shared_ptr<T>(mem, [&q](T *ptr) { sycl::free(ptr, q); }));
+}
+
+constexpr int BL1 = 1;
+constexpr int BL2 = 2;
+constexpr int BL3 = 3;
+
+struct VectorMADIP {
+  register_map_mmhost(BL1,  // buffer_location or aspace
+                      28,   // address width
+                      64,   // data width
+                      16,   // ! latency, must be atleast 16
+                      1,    // read_write_mode, 0: ReadWrite, 1: Read, 2: Write
+                      1,    // maxburst
+                      0,    // align, 0 defaults to alignment of the type
+                      1     // waitrequest, 0: false, 1: true
+                      ) int *x;
+  register_map_mmhost(BL2,  // buffer_location or aspace
+                      28,   // address width
+                      64,   // data width
+                      16,   // ! latency, must be atleast 16
+                      0,    // read_write_mode, 0: ReadWrite, 1: Read, 2: Write
+                      1,    // maxburst
+                      0,    // align, 0 defaults to alignment of the type
+                      1     // waitrequest, 0: false, 1: true
+                      ) int *y;
+  register_map_mmhost(BL3,  // buffer_location or aspace
+                      28,   // address width
+                      64,   // data width
+                      16,   // ! latency, must be atleast 16
+                      0,    // read_write_mode, 0: ReadWrite, 1: Read, 2: Write
+                      1,    // maxburst
+                      0,    // align, 0 defaults to alignment of the type
+                      1     // waitrequest, 0: false, 1: true
+                      ) int *z;
+  int size;
+
+  VectorMADIP(int *x_, int *y_, int *z_, int size_)
+      : x(x_), y(y_), z(z_), size(size_) {}
+
+  register_map_interface void operator()() const {
+    for (int i = 0; i < size; ++i) {
+      int mul = x[i] * y[i];
+      int add = mul + z[i];
+      y[i] = mul;
+      z[i] = add;
+    }
+  }
+};
+
+int main(void) {
+#if defined(FPGA_EMULATOR)
+  ext::intel::fpga_emulator_selector device_selector;
+#else
+  ext::intel::fpga_selector device_selector;
+#endif
+
+  try {
+    sycl::queue q(device_selector, dpc_common::exception_handler,
+                  property::queue::enable_profiling{});
+
+    // Print out the device information.
+    std::cout << "Running on device: "
+              << q.get_device().get_info<info::device::name>() << "\n";
+
+    int size = 15;
+
+    auto x = make_malloc_shared<int>(q, size, BL1);
```

Comment:
These buffer locations being numbers really are confusing to me :/
Each number has a specific meaning? 1 = host, 2 = device etc? or is it just a different identifier?